### ISSUE TITLE
[String] Update the default behavior of the title() method

### ISSUE THREAD

**Issue by javiereguiluz** [symfony/symfony#33892]
**Symfony version(s) affected**: 5.0

**Description**  
The `String::title()` method doesn't work by default like other projects' `title()` method.

**How to reproduce**  
Consider this sample string: `her ROYAL highness`

This is the output of some projects:

* Golang: `Her Royal Highness`
* Python: `Her Royal Highness`
* Twig: `Her Royal Highness`

However, the String component returns:

* `Her ROYAL highness`

**Possible Solution**  
"Title case" is usually expected to work like this -> "lowercase the entire string" + "uppercase the first letter of each word".

Should we do the same for Symfony? Thanks!

**Comment by nicolas-grekas** [comment]
I don't think so: that reduces the possibilities. You can achieve what you describe by calling lower then title. You cannot achieve the current behavior with that proposed behavior.

If we talk about titles in articles, I don't want XML to become Xml, so the current behavior is what makes the most sense to me.

**Comment by yceruto** [comment]
Javier, maybe you're missing the `$allWords` argument of the `title()` method?

**Comment by yceruto** [comment]
Nicolas, maybe it is better default `true` for that `$allWords` argument?

**Comment by nicolas-grekas** [comment]
@yceruto I don't think so, that's why I made it this way :P

**Comment by ostrolucky** [comment]
Uppercasing each word is a bad habit only programmers acquire. This is not a proper grammar. This is not what we learned in schools, this is not what we can see in newspapers. Please stop doing that. Only first letter, or highlights are supposed to be capitalized

**Comment by ro0NL** [comment]
> Uppercasing first letter of each word is a bad habit only programmers acquire.

i tend to disagree, title case (with each word uppercase) exists in general

**Comment by ostrolucky** [comment]
It usually exists on websites thanks to programmers doing that. Nobody writes such sentences on purpose.

**Comment by javiereguiluz** [comment]
@ostrolucky a while ago I thought that too ... but I was completely wrong! @weaverryan taught me that in some parts of the world (USA for example) this is the common case used by all newspapers, magazines and the best websites. It's not a programmer thing at all. IT's related to literature and journalism. Not only it's not bad grammar, but it's the most correct grammar possible.

**Comment by Crell** [comment]
Uppercase Every Word is called "title case" because in *some* conventions in English, that is how you style the title of an article or book.  It has fallen out of favor in recent decades in favor of "Sentence case" for article titles to sound less pretentious, but it got the name "Title Case" precisely because it used to be the standard way to write titles, long before there were programmers or websites.

**Comment by ro0NL** [comment]
@nicolas-grekas can you clarify on picking the default? https://github.com/symfony/symfony/issues/33892#issuecomment-539084141

currently it seems to behave like sentence case actually, see also https://en.wikipedia.org/wiki/Capitalization#Title_case which explicitly mentions "all words" as well.

**Comment by ostrolucky** [comment]
I stand corrected. Seems to me we should have two functions instead then, `title` and `sentence`, with no arguments. Actually though I would prefer all functions to have `Case` suffix, so `title` -> `titleCase`. Not sure title case is used for titles outside the US, but that's how is this casing called

**Comment by konietzny** [comment]
Reinvent the wheel? :D 
IMHO a "Title Case" definition is capitalize every word and not only the first letter. So i prefer to add a `String::sentence()` method to do that.
Also i prefer a `$ignoreUpper` argument instead of `$allWords` to not convert XML to Xml for both methods.

**Comment by nicolas-grekas** [comment]
In Unicode, title case applies to single letter, and is different from either uppercase and lowercase.
See e.g https://www.compart.com/en/unicode/category/Lt

The concept of words/sentence has it's own technical report: https://unicode.org/reports/tr29/ and implementation in ICU (https://www.php.net/manual/en/class.intlbreakiterator.php)

In some languages "title case" should not capitalize the first letter of stop words (e.g. in French).

TL;DR, unless you all are familiar with these concepts (I'm not), this is a lot harder than you think.

We're not looking for a magic formula that will hide all this complexity, because it doesn't exist.
What we're looking for here is for the most useful technical primitive that allows building interesting things on top.

The most useful primitive here is the one that sticks as close as possible to Unicode: one that can "title case", using the Unicode mappings, a single letter. You can then easily combine this primitive with whatever text segmentation logic applies to you. Passing `true` means opting into the most generic (and American centric) way or applying title casing to a list of words, but the algorithm is wrong in many languages and in many situations. That's what would make it a bad default to me.

**Comment by Pierstoval** [comment]
I'm also in for not lowering case before "titleing".

However, I'm also in to having `titleCase()` and `sentenceCase()` differenciation without any argument. In PHP we already have `ucfirst()` and `ucwords()`, and as there are different cultural cases in which both behaviors would be used, I think it's much better to have a really clear wording, and it avoids an `if ($allwords)` in all implementations

**Comment by javiereguiluz** [comment]
I'm closing this because it looks like there is no consensus about what to do. Also, the current `title()` method does the "sentence case" and `title(true)` does the real "title case", so both cases are already covered.

Thank you all for the discussion.
